main/src/library/Fixpoint3/Phase/IndexSelection.flix

```diff
@@ -0,0 +1,387 @@
+    ///
+    @Internal
+    pub def indexProgram(program: RamProgram): RamProgram = match program {
+        case RamProgram.Program(stmt, facts, meta, _) =>
+            // This type annotation is apparently needed?
+            let (newIndexes, newStmt): (Map[RelSym, Vector[Search]], RamStmt) = region rc {
+                let indexes = MutMap.empty(rc);
+                let collectedStmt = collectSearchesStmt(indexes, rc, stmt);
+                let newIndexes =
+                    indexes |>
+                    MutMap.toMap |>
+                    Map.map(AutomaticIndexSelection.minIndex) |>
+                    Map.map(List.toVector);
+                (newIndexes, collectedStmt)
+            };
+            // Expand the searches to full arity
+            let indexesFull =
+                newIndexes |>
+                Map.mapWithKey(relSym -> searches ->
+                    Vector.map(search ->
+                        Vector.range(0, arityOfNonLat(relSym)) |>
+                        Vector.filter(x -> not Vector.memberOf(x, search)) |>
+                        Vector.append(search),
+                        searches
+                    )
+                );
+            // Account for relations that have no body atoms with bound variables
+            let indexesMissing =
+                relSymsOfProgram(program) |>
+                List.filter(relSym -> not Map.memberOf(relSym, indexesFull)) |>
+                List.foldLeft(acc -> relSym -> {
+                    let deltaRelSym = relSymFromPredType(relSym, PredType.Delta, meta);
+                    let search = match relSymToPredType(relSym, meta) {
+                        case PredType.New =>
+                            match Map.get(deltaRelSym, indexesFull) {
+                                case Some(v) => Vector.get(0, v)
+                                case None => Vector.range(0, arityOfNonLat(deltaRelSym))
+                            }
+                        case _ => Vector.range(0, arityOfNonLat(relSym))
+                    };
+                    Map.insert(relSym, Vector#{search}, acc)
+                }, Map#{});
+            let indexes = Map.union(indexesFull, indexesMissing);
+            let (indexedStmt, indexPlacement) = region rc2 {
+                let indexState = UniqueInts.empty(rc2);
+                let indexedStmt = indexStmt(newStmt, indexes, indexState);
+                Map.forEach(relSym -> _ -> {
+                    UniqueInts.getIndex((relSym, 0), indexState);
+                    ()
+                }, indexes);
+                let finalState = UniqueInts.toMap(indexState);
+                (indexedStmt, finalState)
+            };
+            RamProgram.Program(indexedStmt, facts, meta, (indexes, indexPlacement))
+    }
+
+    ///
+    /// Propagates the appropriate indexes to `Query` ops.
+    ///
+    def indexStmt(stmt: RamStmt, constructedIndexes: Indexes, indexState: UniqueInts[(RelSym, Int32), r]): RamStmt \ r = match stmt {
```

Comment:
Here the flix effect system is not powerful enough as documentation.

I _think_ you are returning a new RamStmt while only accessing mutable data in indexState. But I cannot really know.

It would be good if you write was is returned, e.g. "Returns a new RamStmt where all ..."